lib/collection/src/shards/transfer/shard_transfer.rs

```diff
@@ -773,67 +759,64 @@ where
-                Err(error) => {
-                    // Revert queue proxy if we still have any to clean up or prepare for the next attempt
-                    if let Some(replica_set) =
-                        shards_holder.read().await.get_shard(&transfer.shard_id)
-                    {
-                        replica_set.revert_queue_proxy_local().await;
-                    }
-
-                    if matches!(error, CollectionError::Cancelled { .. }) {
-                        return false;
-                    }
-                    log::error!(
-                        "Failed to transfer shard {} -> {}: {error}",
+        let mut result = Err(cancel::Error::Cancelled);
+
+        for attempt in 0..MAX_RETRY_COUNT {
+            let future = async {
+                if attempt > 0 {
+                    sleep(RETRY_DELAY * attempt as u32).await;
+
+                    log::warn!(
+                        "Retrying shard transfer {collection_id}:{} -> {} (retry {attempt})",
                         transfer.shard_id,
                         transfer.to,
                     );
-
-                    false
                 }
+
+                transfer_shard(
+                    transfer.clone(),
+                    shards_holder.clone(),
+                    consensus.as_ref(),
+                    collection_id.clone(),
+                    &collection_name,
+                    channel_service.clone(),
+                    &snapshots_path,
+                    &temp_dir,
+                )
+                .await
             };
-            if cancel.is_cancelled() {
-                return false;
-            }
-            if !finished {
-                tries -= 1;
-                log::warn!(
-                    "Retrying shard transfer {} -> {} (retry {})",
+
+            result = cancel::future::cancel_on_token(cancel.clone(), future).await;
+
+            // If `transfer_shard` returned error...
+            if let Ok(Err(err)) = &result {
+                log::error!(
+                    "Failed to transfer shard {collection_id}:{} -> {}: {err}",
                     transfer.shard_id,
                     transfer.to,
-                    MAX_RETRY_COUNT - tries
                 );
-                let exp_timeout = RETRY_DELAY * (MAX_RETRY_COUNT - tries) as u32;
-                sleep(exp_timeout).await;
+            }
+
+            // If `transfer_shard` returned error or task was cancelled...
+            if matches!(result, Ok(Err(_)) | Err(_)) {
+                // Revert queue proxy if we still have any to prepare for the next attempt
+                if let Some(shard) = shards_holder.read().await.get_shard(&transfer.shard_id) {
+                    shard.revert_queue_proxy_local().await;
+                }
+            }
+
+            // If `transfer_shard` returned success or task was cancelled...
+            if matches!(result, Ok(Ok(())) | Err(_)) {
+                break;
             }
```

Comment:
What about this? We use this pattern in some other places as well.

```suggestion
            let is_cancelled = result.is_err();
            let is_ok = matches!(result, Ok(Ok(())));
            let is_err = matches!(result, Ok(Err(_)));

            // If `transfer_shard` returned error or task was cancelled...
            if is_cancelled || is_err {
                // Revert queue proxy if we still have any to prepare for the next attempt
                if let Some(shard) = shards_holder.read().await.get_shard(&transfer.shard_id) {
                    shard.revert_queue_proxy_local().await;
                }
            }

            // If `transfer_shard` returned success or task was cancelled...
            if is_ok || is_cancelled {
                break;
            }
```

Reply:
That's... much better, actually. I don't know why I didn't implement it like that in the first place. 🤦‍♀️

Reply:
Done.

Reply:
I also always get blind for improvements like this after refactoring some code a million times. :laughing: 

> Done.

Awesome!